### ISSUE TITLE
Optional values missing in Record causes ValidationError.

### ISSUE THREAD

**Issue by amrx101** [flavray/avro-rs#106]
fn main() -> Result<(), Error> {
        let raw_schema = r#"
            {
                "type": "record",
                "name": "test",
                "fields": [
                    {"name": "a", "type": "string"},
                    {"name": "b", "type": "string"},
                    {"name": "c",  "type": ["null", "string"]}
                ]
            }
        "#;

        let schema = Schema::parse_str(raw_schema)?;

        println!("==============");
        println!("{:?}", schema);

        let mut writer = Writer::with_codec(&schema, Vec::new(), Codec::Deflate);

        let mut record = Record::new(writer.schema()).unwrap();
        let mut check : BTreeMap<String, String> = BTreeMap::new();
        check.insert("a".to_owned(), "100".to_owned());
        check.insert("b".to_owned(), "200".to_owned());
        for (key, value) in check.iter(){
            record.put(key, value.clone());
        }

        writer.append(record)?;

        // let test = Test {
        //     a: "27".to_owned(),
        //     b: "foo".to_owned(),
        //     c: None,
        // };

        // writer.append_ser(test)?;

        writer.flush()?;

        let input = writer.into_inner();
        let reader = Reader::with_schema(&schema, &input[..])?;

        for record in reader {
            println!("{:?}", from_value::<Test>(&record?));
        }
    Ok(())
}

This causes ValidationError for non matching schema.

If Record being written doesn't have a key whose value is `optional` the `Writer` should infer that and not cause a validation error, right?

**Comment by poros** [comment]
Two things at play here.

One is a bug that has been fixed in master ( #103 ) but that I have yet to release. I was waiting for another PR to land, but I'll just release a patch version now since it is bothering you.

Although, I think that the avro spec and your expectations do not exactly match. You can read more about unions at https://avro.apache.org/docs/current/spec.html#Unions and about default values at 
https://avro.apache.org/docs/current/spec.html#schema_record and https://avro.apache.org/docs/current/spec.html#Schema+Resolution.

In practice, these are the sentences that explain how default values are handled in avro for record fields:
> default: A default value for this field, used when reading instances that lack this field (optional).

>if the writer's record contains a field with a name not present in the reader's record, the writer's value for that field is ignored.
> if the reader's record schema has a field that contains a default value, and writer's schema does not have a field with the same name, then the reader should use the default value from its field.
> if the reader's record schema has a field with no default value, and writer's schema does not have a field with the same name, an error is signalled.

Practically, this means that:

-  unions with null do *not* explicitly imply a null default value

- default values are only used at the reader side, not at the writer side

A working example (with current master, *not* version 0.6.6) would be:
```
    #[test]
    fn test_default() -> Result<(), Error> {

        #[derive(Debug, Serialize, Deserialize)]
        struct Test {
            a: String,
            b: String,
            c: Option<String>,
        }

        let raw_schema = r#"
        {
            "type": "record",
            "name": "test",
            "fields": [
                {"name": "a", "type": "string"},
                {"name": "b", "type": "string"}
            ]
        }
        "#;

        let raw_schema_r = r#"
        {
            "type": "record",
            "name": "test",
            "fields": [
                {"name": "a", "type": "string"},
                {"name": "b", "type": "string"},
                {"name": "c",  "type": ["null", "string"], "default": "null"}
            ]
        }"#;

        let schema = Schema::parse_str(raw_schema)?;
        let schema_r = Schema::parse_str(raw_schema_r)?;

        println!("==============");
        println!("{:?}", schema);

        let mut writer = Writer::with_codec(&schema, Vec::new(), Codec::Deflate);

        let mut record = Record::new(writer.schema()).unwrap();
        record.put("a", "antonio".to_owned());
        record.put("b", "antonio".to_owned());

        writer.append(record)?;

        writer.flush()?;

        let input = writer.into_inner();
        let reader = Reader::with_schema(&schema_r, &input[..])?;

        for record in reader {
            println!("{:?}", from_value::<Test>(&record?));
        }
        Ok(())
```

I hope this (and version 0.6.7) will answer your question, but I'll leave this issue open for a bit before closing it as "won't fix", in case I misunderstood what you wanted to ask.

**Comment by amrx101** [comment]
Sorry for really late response.

1. 103 actually the union resolution does solve my issue.

2. As per the mismatch of schema between Reader and Writer doesn't really apply for me as I control the both ends of the schema.

**Comment by amrx101** [comment]
@poros , in which version is this available or should I use master itself?

**Comment by poros** [comment]
I recommend you use the latest version, but it should be available from version 0.7.0.

**Comment by poros** [comment]
As it isn't in the avro spec (at least, as far as I can understand the spec, which is indeed a quite contrived at times), I don't plan on supporting nullable values on the writer side any time soon.
